src/Microsoft.AspNetCore.Http.Connections/Internal/Transports/HttpUpgradeTransport.cs

```diff
@@ -0,0 +1,279 @@
+            var receiving = StartReceiving(socket);
+            var sending = StartSending(socket);
+
+            // Wait for send or receive to complete
+            var trigger = await Task.WhenAny(receiving, sending);
+
+            if (trigger == receiving)
+            {
+                Log.WaitingForSend(_logger);
+
+                // We're waiting for the application to finish and there are 2 things it could be doing
+                // 1. Waiting for application data
+                // 2. Waiting for a websocket send to complete
+
+                // Cancel the application so that ReadAsync yields
+                _application.Input.CancelPendingRead();
+
+                using (var delayCts = new CancellationTokenSource())
+                {
+                    var resultTask = await Task.WhenAny(sending, Task.Delay(_options.CloseTimeout, delayCts.Token));
+
+                    if (resultTask != sending)
+                    {
+                        // We timed out so now we're in ungraceful shutdown mode
+                        Log.CloseTimedOut(_logger);
+
+                        // Abort the websocket if we're stuck in a pending send to the client
+                        _aborted = true;
+
+                        context.Abort();
+                    }
+                    else
+                    {
+                        delayCts.Cancel();
+                    }
+                }
+            }
+            else
+            {
+                Log.WaitingForClose(_logger);
+
+                // We're waiting on the websocket to close and there are 2 things it could be doing
+                // 1. Waiting for websocket data
+                // 2. Waiting on a flush to complete (backpressure being applied)
+
+                using (var delayCts = new CancellationTokenSource())
+                {
+                    var resultTask = await Task.WhenAny(receiving, Task.Delay(_options.CloseTimeout, delayCts.Token));
+
+                    if (resultTask != receiving)
+                    {
+                        // Abort the websocket if we're stuck in a pending receive from the client
+                        _aborted = true;
+
+                        context.Abort();
+
+                        // Cancel any pending flush so that we can quit
+                        _application.Output.CancelPendingFlush();
+                    }
+                    else
+                    {
+                        delayCts.Cancel();
+                    }
+                }
+            }
+        }
+
+        private async Task StartReceiving(Stream stream)
+        {
+            try
+            {
+                while (true)
+                {
+#if NETCOREAPP2_2
+                    // Do a 0 byte read so that idle connections don't allocate a buffer when waiting for a read
+                    await stream.ReadAsync(Memory<byte>.Empty, CancellationToken.None);
```

Comment:
I think the issue is, you do a 0-byte read here and there was a 0 byte message from the pipe (the close), then when you do the actual read below you wait for a message that was already read so it will "hang"